### PR TITLE
use memory allocator in byte array dictionary

### PR DESCRIPTION
**Pull request by achille-roussel**
Indexing BYTE_ARRAY columns is a frequent use case and when done at high throughput, the creation of string copies to capture the dictionary keys would result in excessive time spent in memory allocations.

This PR addresses the issue by using a local allocator in the `byteArrayDictionary` type to amortize the cost of memory allocations for the dictionary keys.

### CODE REVIEW

File: allocator.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	b := a.makeBytes(len(v))
+	copy(b, v)
+	return unsafecast.BytesToString(b)
+}
+
+func (a *allocator) reset() {
+	a.buffer = a.buffer[:0]
+}
+
+// rowAllocator is a memory allocator used to make copy of rows referencing
```
Comment (kevinburkesegment) [inline allocator.go:41]:
```suggestion
// rowAllocator is a memory allocator used to make a copy of rows referencing
```

File: allocator.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+// memory buffers that parquet-go does not have ownership of.
+//
+// This type is used in the implementation of a various readers and writers that
```
Comment (kevinburkesegment) [inline allocator.go:44]:
```suggestion
// This type is used in the implementation of various readers and writers that
```

File: allocator.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+// need to capture rows passed to the ReadRows/WriteRows methods.
+//
+// See: RowBuffer, DedupeRowReader, DedupeRowWriter
+type rowAllocator struct{ allocator }
+
+func (a *rowAllocator) capture(row Row) {
+	for i, v := range row {
+		switch kind := v.Kind(); kind {
+		case ByteArray, FixedLenByteArray:
+			row[i].ptr = unsafecast.AddressOfBytes(a.copyBytes(v.byteArray()))
```
Comment (kevinburkesegment) [inline allocator.go:54]: default case?

File: allocator.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	copy(b, v)
+	return unsafecast.BytesToString(b)
+}
+
+func (a *allocator) reset() {
+	a.buffer = a.buffer[:0]
+}
+
+// rowAllocator is a memory allocator used to make copy of rows referencing
+// memory buffers that parquet-go does not have ownership of.
```
Comment (kevinburkesegment) [inline allocator.go:42]: I'd appreciate another sentence about why we can't just use the Go allocator for this

File: allocator.go (hunk 3 above)
Comment (achille-roussel) [reply allocator.go:54]: There's nothing to do for the default case 👍

File: allocator.go (hunk 4 above)
Comment (achille-roussel) [reply allocator.go:42]: Let me know how that reads https://github.com/segmentio/parquet-go/pull/445/commits/d2f5a1f9ca5dbfc8a553c8a535a09d00ffc54c87

**Review by kevinburkesegment — APPROVED** [review]

**Pull request merged**